### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by sameedmohsin28**
In this PR, I have

- Created model classes for all entities.
- Set up associations between models.
- Created additional custom methods in models.

### CODE REVIEW

File: app/models/like.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (franclobo) [inline app/models/like.rb:2]: - [OPTIONAL] As you have done in the post the foreign key is author_id, so you join it, Please 🙏🏽  correct this line to join the  `like` table to the correct column 👇🏽 

```ruby
belongs_to :author, class_name: 'User'
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
Comment (franclobo) [inline app/models/comment.rb:2]: - [OPTIONAL] As you have done in the post the foreign key is author_id, so you join it, Please 🙏🏽  correct this line to join the  `like` table to the correct column 👇🏽 

```ruby
belongs_to :author, class_name: 'User'
```

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: :author_id
+  has_many :likes
+  has_many :comments
```
Comment (franclobo) [inline app/models/user.rb]: - [ ] Please 🙏🏽 add the corresponding foreing key to the likes and comments tables as you did for the posts table.

File: app/models/post.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (franclobo) [inline app/models/post.rb]: - [ ] You have declared the foreign key in the user modal so here is not necessary. Please remove it to be DRY.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+
+  def most_recent
+    posts.order(created_at: :desc).limit(3)
+  end
```
Comment (franclobo) [inline app/models/user.rb:8]: - **[OPTIONAL]** We can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to database, then make the `update_post_counter` method privateStudents can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the #update_post_counter method every time a new post has been saved to database, then make the update_post_counter method private.

```ruby
after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/like.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  belongs_to :post
+
+  def update_likes_counter_for_post
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (franclobo) [inline app/models/like.rb:10]: - **[OPTIONAL]** Just like we have in the user modal, `after_save :update_comments_counter`

File: app/models/comment.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+  belongs_to :post
+
+  def update_comments_counter_for_post
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (franclobo) [inline app/models/comment.rb]: - **[OPTIONAL]** Just like we have in the user modal, `after_save :update_comments_counter`

**Review by franclobo — CHANGES_REQUESTED** [review]
Hi @sameedmohsin28, 

# Required Changes ♻️

so-close-jim-halpert

Good job so far!
You are doing it well 👏🏽 there are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

- No linter errors ✔️ 
- Descriptive PR ✔️ 
- Correct Git flow ✔️ 
- Correct models ✔️ 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You should consider this:

## - The user table should have a unique primary key that is named author_id for the post table, in the post table it will be the foreign key, so it doesn't make sense to have another column user_id to the likes and comments tables, does it?

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/like.rb (hunk 1 above)
Comment (sameedmohsin28) [reply app/models/like.rb:2]: Hi @franclobo ,
Thanks for your kind, humble, and detailed review. I appreciate the efforts you have tried putting in to explain your suggestions. However, I request you to kindly think about your suggestions as I think the changes you suggested are not mentioned in the project requirements. If I am mistaken, kindly explain to me further as to how am I mistaken. Along with this comment, I am writing two other comments below so please check them too.

As mentioned in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png), the `author` is used only for `Post`. The `Like` has the attribute which goes by the name `UserID`.

File: app/models/comment.rb (hunk 2 above)
Comment (sameedmohsin28) [reply app/models/comment.rb:2]: As mentioned in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png), the `author` is used only for `Post`. The `Comment` has the attribute which goes by the name `UserID`.

File: app/models/user.rb (hunk 3 above)
Comment (sameedmohsin28) [reply app/models/user.rb]: If what I said above is correct, then I guess this change too does not need to be made.

File: app/models/like.rb (hunk 1 above)
Comment (franclobo) [reply app/models/like.rb:2]: Hello, @sameedmohsin28.

You have the reason in your point, as CR I have been reviewing according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). I hope the people in charge can change to use the same ERD diagram. For this reason I have changed the required change to OPTIONAL.

Thanks for your suggestion.

File: app/models/comment.rb (hunk 2 above)
Comment (franclobo) [reply app/models/comment.rb:2]: Hello, @sameedmohsin28.

You have the reason in your point, as CR I have been reviewing according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). I hope the people in charge can change to use the same ERD diagram. For this reason I have changed the required change to OPTIONAL.

Thanks for your suggestion.

File: app/models/user.rb (hunk 3 above)
Comment (franclobo) [reply app/models/user.rb]: Hello, @sameedmohsin28.

You have the reason in your point, as CR I have been reviewing according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png). I hope the people in charge can change to use the same ERD diagram. For this reason I rephrased the required change.

Thanks for your suggestion.

Comment (sameedmohsin28) [reply app/models/user.rb]: @franclobo 

As per [Active Record lesson by Microverse](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/ActiveRecord_lesson.md#associations), and [Ruby on Rails guide](https://guides.rubyonrails.org/association_basics.html#the-types-of-associations), adding the corresponding foreign key is only required when the name of the foreign key column is set something other than the default name.
Kindly consider this as well. Thanks!

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @sameedmohsin28 , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  All changes resolved :+1: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**